In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()
import warnings
warnings.filterwarnings('ignore')

In [2]:
df = pd.read_csv('WA_Fn-UseC_-HR-Employee-Attrition.csv')

In [3]:
pd.set_option('display.max_columns',None)

In [4]:
df.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   EnvironmentSatisfaction  Gender  HourlyRate  JobInvolvement  JobLevel  \
0                        2  Female          94               3         2   
1                        3    Male          61               2         2   
2                        4    Male          92               2         1   
3                        4  Female          56               3         1   
4                        1    Male          40               3         1   

                 JobRole  JobSatisfaction MaritalStatus  MonthlyIncome  \
0        Sales Executive                4        Single           5993   
1     Research Scientist                2       Married           5130   
2  Laboratory Technician                3        Single           2090   
3     Research Scientist                3       Married           2909   
4  Laboratory Technician                2       Married           3468   

   MonthlyRate  NumCompaniesWorked Over18 OverTime  PercentSalaryHike  \
0        19479                   8      Y      Yes                 11   
1        24907                   1      Y       No                 23   
2         2396                   6      Y      Yes                 15   
3        23159                   1      Y      Yes                 11   
4        16632                   9      Y       No                 12   

   PerformanceRating  RelationshipSatisfaction  StandardHours  \
0                  3                         1             80   
1                  4                         4             80   
2                  3                         2             80   
3                  3                         3             80   
4                  3                         4             80   

   StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                 0                  8                      0   
1                 1                 10                      3   
2                 0                  7                      3   
3                 0                  8                      3   
4                 1                  6                      3   

   WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
0                1               6                   4   
1                3              10                   7   
2                3               0                   0   
3                3               8                   7   
4                3               2                   2   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     5  
1                        1                     7  
2                        0                     0  
3                        3                     0  
4                        2                     2

In [5]:
df.shape

(1470, 35)

In [6]:
df.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'DailyRate', 'Department',
       'DistanceFromHome', 'Education', 'EducationField', 'EmployeeCount',
       'EmployeeNumber', 'EnvironmentSatisfaction', 'Gender', 'HourlyRate',
       'JobInvolvement', 'JobLevel', 'JobRole', 'JobSatisfaction',
       'MaritalStatus', 'MonthlyIncome', 'MonthlyRate', 'NumCompaniesWorked',
       'Over18', 'OverTime', 'PercentSalaryHike', 'PerformanceRating',
       'RelationshipSatisfaction', 'StandardHours', 'StockOptionLevel',
       'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance',
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'YearsWithCurrManager'],
      dtype='object')

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

In [8]:
target = 'Attrition'
indep_cols = df.columns[df.columns != 'Attrition']

In [164]:
# Description of numerical columns
df.describe()

Age    DailyRate  DistanceFromHome    Education  EmployeeCount  \
count  1470.000000  1470.000000       1470.000000  1470.000000         1470.0   
mean     36.923810   802.485714          9.192517     2.912925            1.0   
std       9.135373   403.509100          8.106864     1.024165            0.0   
min      18.000000   102.000000          1.000000     1.000000            1.0   
25%      30.000000   465.000000          2.000000     2.000000            1.0   
50%      36.000000   802.000000          7.000000     3.000000            1.0   
75%      43.000000  1157.000000         14.000000     4.000000            1.0   
max      60.000000  1499.000000         29.000000     5.000000            1.0   

       EmployeeNumber  EnvironmentSatisfaction   HourlyRate  JobInvolvement  \
count     1470.000000              1470.000000  1470.000000     1470.000000   
mean      1024.865306                 2.721769    65.891156        2.729932   
std        602.024335                 1.093082    20.329428        0.711561   
min          1.000000                 1.000000    30.000000        1.000000   
25%        491.250000                 2.000000    48.000000        2.000000   
50%       1020.500000                 3.000000    66.000000        3.000000   
75%       1555.750000                 4.000000    83.750000        3.000000   
max       2068.000000                 4.000000   100.000000        4.000000   

          JobLevel  JobSatisfaction  MonthlyIncome   MonthlyRate  \
count  1470.000000      1470.000000    1470.000000   1470.000000   
mean      2.063946         2.728571    6502.931293  14313.103401   
std       1.106940         1.102846    4707.956783   7117.786044   
min       1.000000         1.000000    1009.000000   2094.000000   
25%       1.000000         2.000000    2911.000000   8047.000000   
50%       2.000000         3.000000    4919.000000  14235.500000   
75%       3.000000         4.000000    8379.000000  20461.500000   
max       5.000000         4.000000   19999.000000  26999.000000   

       NumCompaniesWorked  PercentSalaryHike  PerformanceRating  \
count         1470.000000        1470.000000        1470.000000   
mean             2.693197          15.209524           3.153741   
std              2.498009           3.659938           0.360824   
min              0.000000          11.000000           3.000000   
25%              1.000000          12.000000           3.000000   
50%              2.000000          14.000000           3.000000   
75%              4.000000          18.000000           3.000000   
max              9.000000          25.000000           4.000000   

       RelationshipSatisfaction  StandardHours  StockOptionLevel  \
count               1470.000000         1470.0       1470.000000   
mean                   2.712245           80.0          0.793878   
std                    1.081209            0.0          0.852077   
min                    1.000000           80.0          0.000000   
25%                    2.000000           80.0          0.000000   
50%                    3.000000           80.0          1.000000   
75%                    4.000000           80.0          1.000000   
max                    4.000000           80.0          3.000000   

       TotalWorkingYears  TrainingTimesLastYear  WorkLifeBalance  \
count        1470.000000            1470.000000      1470.000000   
mean           11.279592               2.799320         2.761224   
std             7.780782               1.289271         0.706476   
min             0.000000               0.000000         1.000000   
25%             6.000000               2.000000         2.000000   
50%            10.000000               3.000000         3.000000   
75%            15.000000               3.000000         3.000000   
max            40.000000               6.000000         4.000000   

       YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
count     1470.000000         1470.000000              1470.

In [165]:
numerical_cols = df.describe().columns

In [166]:
numerical_cols

Index(['Age', 'DailyRate', 'DistanceFromHome', 'Education', 'EmployeeCount',
       'EmployeeNumber', 'EnvironmentSatisfaction', 'HourlyRate',
       'JobInvolvement', 'JobLevel', 'JobSatisfaction', 'MonthlyIncome',
       'MonthlyRate', 'NumCompaniesWorked', 'PercentSalaryHike',
       'PerformanceRating', 'RelationshipSatisfaction', 'StandardHours',
       'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear',
       'WorkLifeBalance', 'YearsAtCompany', 'YearsInCurrentRole',
       'YearsSinceLastPromotion', 'YearsWithCurrManager'],
      dtype='object')

In [167]:
df.describe(include = 'object')

Attrition BusinessTravel              Department EducationField Gender  \
count       1470           1470                    1470           1470   1470   
unique         2              3                       3              6      2   
top           No  Travel_Rarely  Research & Development  Life Sciences   Male   
freq        1233           1043                     961            606    882   

                JobRole MaritalStatus Over18 OverTime  
count              1470          1470   1470     1470  
unique                9             3      1        2  
top     Sales Executive       Married      Y       No  
freq                326           673   1470     1054

In [168]:
cat_cols = df.describe(include = 'object').columns[df.describe(include = 'object').columns != 'Attrition']

In [169]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import OrdinalEncoder
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import LabelEncoder
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression

In [170]:
# Since the Over18 column has only one unique value , we are going to drop it.
df.drop(['Over18'],axis = 1,inplace = True)

In [171]:
cat_cols = cat_cols.to_list()
cat_cols

['BusinessTravel',
 'Department',
 'EducationField',
 'Gender',
 'JobRole',
 'MaritalStatus',
 'Over18',
 'OverTime']

In [172]:
cat_cols.remove('Over18')

In [173]:
cat_cols

['BusinessTravel',
 'Department',
 'EducationField',
 'Gender',
 'JobRole',
 'MaritalStatus',
 'OverTime']

In [174]:
df.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'DailyRate', 'Department',
       'DistanceFromHome', 'Education', 'EducationField', 'EmployeeCount',
       'EmployeeNumber', 'EnvironmentSatisfaction', 'Gender', 'HourlyRate',
       'JobInvolvement', 'JobLevel', 'JobRole', 'JobSatisfaction',
       'MaritalStatus', 'MonthlyIncome', 'MonthlyRate', 'NumCompaniesWorked',
       'OverTime', 'PercentSalaryHike', 'PerformanceRating',
       'RelationshipSatisfaction', 'StandardHours', 'StockOptionLevel',
       'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance',
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'YearsWithCurrManager'],
      dtype='object')

In [175]:
df[cat_cols]

BusinessTravel              Department EducationField  Gender  \
0         Travel_Rarely                   Sales  Life Sciences  Female   
1     Travel_Frequently  Research & Development  Life Sciences    Male   
2         Travel_Rarely  Research & Development          Other    Male   
3     Travel_Frequently  Research & Development  Life Sciences  Female   
4         Travel_Rarely  Research & Development        Medical    Male   
...                 ...                     ...            ...     ...   
1465  Travel_Frequently  Research & Development        Medical    Male   
1466      Travel_Rarely  Research & Development        Medical    Male   
1467      Travel_Rarely  Research & Development  Life Sciences    Male   
1468  Travel_Frequently                   Sales        Medical    Male   
1469      Travel_Rarely  Research & Development        Medical    Male   

                        JobRole MaritalStatus OverTime  
0               Sales Executive        Single      Yes  
1            Research Scientist       Married       No  
2         Laboratory Technician        Single      Yes  
3            Research Scientist       Married      Yes  
4         Laboratory Technician       Married       No  
...                         ...           ...      ...  
1465      Laboratory Technician       Married       No  
1466  Healthcare Representative       Married       No  
1467     Manufacturing Director       Married      Yes  
1468            Sales Executive       Married       No  
1469      Laboratory Technician       Married       No  

[1470 rows x 7 columns]

In [176]:
for col in cat_cols:
    print(col)
    print(df[col].unique())
    print(" ")

BusinessTravel
['Travel_Rarely' 'Travel_Frequently' 'Non-Travel']
 
Department
['Sales' 'Research & Development' 'Human Resources']
 
EducationField
['Life Sciences' 'Other' 'Medical' 'Marketing' 'Technical Degree'
 'Human Resources']
 
Gender
['Female' 'Male']
 
JobRole
['Sales Executive' 'Research Scientist' 'Laboratory Technician'
 'Manufacturing Director' 'Healthcare Representative' 'Manager'
 'Sales Representative' 'Research Director' 'Human Resources']
 
MaritalStatus
['Single' 'Married' 'Divorced']
 
OverTime
['Yes' 'No']
 


In [177]:
# Business travel and overtime column will be ordinal encoded and other categorical columns will be one hot encoded.
ordinal_encoder = OrdinalEncoder(categories=[['Non-Travel','Travel_Rarely','Travel_Frequently'],['No','Yes']])

In [178]:
data1 = df[['BusinessTravel','OverTime']]
data1 = ordinal_encoder.fit_transform(data1)

In [179]:
ordinal_encoder.categories_

[array(['Non-Travel', 'Travel_Rarely', 'Travel_Frequently'], dtype=object),
 array(['No', 'Yes'], dtype=object)]

In [180]:
ohe_cols = [col for col in cat_cols if col not in ['BusinessTravel','OverTime']]
ohe_cols

['Department', 'EducationField', 'Gender', 'JobRole', 'MaritalStatus']

In [181]:
data2 = df[ohe_cols]
ohe = OneHotEncoder(drop = 'first',sparse_output= False,dtype = np.int32)
data2 = ohe.fit_transform(data2)

In [182]:
ohe.categories_

[array(['Human Resources', 'Research & Development', 'Sales'], dtype=object),
 array(['Human Resources', 'Life Sciences', 'Marketing', 'Medical',
        'Other', 'Technical Degree'], dtype=object),
 array(['Female', 'Male'], dtype=object),
 array(['Healthcare Representative', 'Human Resources',
        'Laboratory Technician', 'Manager', 'Manufacturing Director',
        'Research Director', 'Research Scientist', 'Sales Executive',
        'Sales Representative'], dtype=object),
 array(['Divorced', 'Married', 'Single'], dtype=object)]

In [183]:
df[numerical_cols]

Age  DailyRate  DistanceFromHome  Education  EmployeeCount  \
0      41       1102                 1          2              1   
1      49        279                 8          1              1   
2      37       1373                 2          2              1   
3      33       1392                 3          4              1   
4      27        591                 2          1              1   
...   ...        ...               ...        ...            ...   
1465   36        884                23          2              1   
1466   39        613                 6          1              1   
1467   27        155                 4          3              1   
1468   49       1023                 2          3              1   
1469   34        628                 8          3              1   

      EmployeeNumber  EnvironmentSatisfaction  HourlyRate  JobInvolvement  \
0                  1                        2          94               3   
1                  2                        3          61               2   
2                  4                        4          92               2   
3                  5                        4          56               3   
4                  7                        1          40               3   
...              ...                      ...         ...             ...   
1465            2061                        3          41               4   
1466            2062                        4          42               2   
1467            2064                        2          87               4   
1468            2065                        4          63               2   
1469            2068                        2          82               4   

      JobLevel  JobSatisfaction  MonthlyIncome  MonthlyRate  \
0            2                4           5993        19479   
1            2                2           5130        24907   
2            1                3           2090         2396   
3            1                3           2909        23159   
4            1                2           3468        16632   
...        ...              ...            ...          ...   
1465         2                4           2571        12290   
1466         3                1           9991        21457   
1467         2                2           6142         5174   
1468         2                2           5390        13243   
1469         2                3           4404        10228   

      NumCompaniesWorked  PercentSalaryHike  PerformanceRating  \
0                      8                 11                  3   
1                      1                 23                  4   
2                      6                 15                  3   
3                      1                 11                  3   
4                      9                 12                  3   
...                  ...                ...                ...   
1465                   4                 17                  3   
1466                   4                 15                  3   
1467                   1                 20                  4   
1468                   2                 14                  3   
1469                   2                 12                  3   

      RelationshipSatisfaction  StandardHours  StockOptionLevel  \
0                            1             80                 0   
1                            4             80                 1   
2                            2             80                 0   
3                            3             80                 0   
4                            4             80                 1   
...                        ...            ...               ...   
1465                         3             80                 1   
1466                         1             80                 1   
1467                         2             80                 1   
1468                         4             8

In [184]:
df['EmployeeCount'].unique()

array([1], dtype=int64)

In [185]:
# EmployeeCount is not meaningful in model creation. The feature is containing only one unique value.We will drop the column.
# Similarily , EmployeeNumber columns is not meaningful in model creation , The feature is for providing a unique identification for each employee.We will drop it
# standardhours column has also only one unique value.

df.drop(['EmployeeCount','EmployeeNumber','StandardHours'],axis = 1 , inplace = True)


In [186]:
# Removing the above two columns from the numerical_cols list.

numerical_cols = numerical_cols.to_list()
numerical_cols.remove('EmployeeNumber')
numerical_cols.remove('EmployeeCount')
numerical_cols.remove('StandardHours')
numerical_cols

['Age',
 'DailyRate',
 'DistanceFromHome',
 'Education',
 'EnvironmentSatisfaction',
 'HourlyRate',
 'JobInvolvement',
 'JobLevel',
 'JobSatisfaction',
 'MonthlyIncome',
 'MonthlyRate',
 'NumCompaniesWorked',
 'PercentSalaryHike',
 'PerformanceRating',
 'RelationshipSatisfaction',
 'StockOptionLevel',
 'TotalWorkingYears',
 'TrainingTimesLastYear',
 'WorkLifeBalance',
 'YearsAtCompany',
 'YearsInCurrentRole',
 'YearsSinceLastPromotion',
 'YearsWithCurrManager']

In [187]:
data3 = df[numerical_cols].values
data3.shape

(1470, 23)

In [188]:
print(data1.shape)
print(data2.shape)

(1470, 2)
(1470, 18)


In [189]:
# Scaling the numerical_cols ie data3 and ordinal encoded features ie data1

from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
data3 = sc.fit_transform(data3)
sc1 = StandardScaler()
data1 = sc1.fit_transform(data1)

In [190]:
# We will create a new matrix by concatenation of data1 , data2 and data3 . This matrix will consist of the independent features for our ml model . data1 is derived from ordinal categorical variables ,data2 is derived from the nominal categorical variables and data3 is derived from the numerical variables.

X = np.concatenate([data1,data2,data3],axis = 1) 

In [191]:
X.shape

(1470, 43)

In [192]:
# Target variable
df[target]

0       Yes
1        No
2       Yes
3        No
4        No
       ... 
1465     No
1466     No
1467     No
1468     No
1469     No
Name: Attrition, Length: 1470, dtype: object

In [193]:
# encoding target variable
le = LabelEncoder()
le.fit(df[target])

LabelEncoder()

In [194]:
le.classes_

array(['No', 'Yes'], dtype=object)

In [195]:
y = le.transform(df[target])

In [196]:
y.shape

(1470,)

In [197]:
# Train Test Split

xtrain,xtest,ytrain,ytest = train_test_split(X,y,test_size = 0.2,stratify=y,random_state= 10)
print(xtrain.shape)
print(ytrain.shape)

(1176, 43)
(1176,)


In [198]:
lr = LogisticRegression()
lr.fit(xtrain,ytrain)
print('Train accuracy - ',lr.score(xtrain,ytrain))
print('Test accuracy - ',lr.score(xtest,ytest))

Train accuracy -  0.8962585034013606
Test accuracy -  0.8843537414965986


In [199]:
# Precision
from sklearn.metrics import accuracy_score,precision_score,recall_score,f1_score

ytest_pred = lr.predict(xtest)
print('Accuracy Score  = ', accuracy_score(y_true = ytest , y_pred = ytest_pred))
print('Precision Score  = ', precision_score(y_true = ytest , y_pred = ytest_pred))
print('Recall Score = ' , recall_score(y_true = ytest , y_pred = ytest_pred))
print('F1 score = ' , f1_score(y_true=ytest , y_pred = ytest_pred))

Accuracy Score  =  0.8843537414965986
Precision Score  =  0.76
Recall Score =  0.40425531914893614
F1 score =  0.5277777777777778


In [200]:
ytest

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0,
       0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0,
       0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 1])

In [201]:
ytest_pred

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0,
       0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 1])

In [203]:
# The data is imbalanced .Most of the observations are having 0 value.

np.unique(y,return_counts=True)

(array([0, 1]), array([1233,  237], dtype=int64))

In [204]:
# SMOTE for balancing the dataset . We can also try with different k_neighbors values.

from imblearn.over_sampling import SMOTE
sm = SMOTE(random_state=10)
xtrain_res , ytrain_res = sm.fit_resample(xtrain,ytrain)

In [205]:
ytrain.shape , ytrain_res.shape,np.unique(ytrain_res,return_counts=True)

((1176,), (1972,), (array([0, 1]), array([986, 986], dtype=int64)))

In [206]:
# Implementation of Logistic regression again.
lr = LogisticRegression()
lr.fit(xtrain_res,ytrain_res)
print('Train accuracy - ',lr.score(xtrain_res,ytrain_res))
print('Test accuracy - ',lr.score(xtest,ytest))

Train accuracy -  0.821501014198783
Test accuracy -  0.7687074829931972


In [207]:
# Checking out other metrics
ytest_pred = lr.predict(xtest)
accuracy = accuracy_score(y_true = ytest , y_pred = ytest_pred)
precision = precision_score(y_true = ytest , y_pred = ytest_pred)
recall = recall_score(y_true = ytest , y_pred = ytest_pred)
f1score = f1_score(y_true=ytest , y_pred = ytest_pred)
print('Accuracy Score  = ', accuracy)
print('Precision Score  = ', precision)
print('Recall Score = ' , recall)
print('F1 score = ' , f1score)

Accuracy Score  =  0.7687074829931972
Precision Score  =  0.3764705882352941
Recall Score =  0.6808510638297872
F1 score =  0.48484848484848486


In [208]:
# creating the confusion matrix
predicted = pd.Series(ytest_pred,name = 'predicted')
actual = pd.Series(ytest , name = 'actual')
pd.crosstab(index = predicted , columns = actual)

# High number of false positives .

actual       0   1
predicted         
0          194  15
1           53  32

In [209]:
results = pd.DataFrame(columns = ['accuracy' , 'precison' ,'recall' ,'f1_score'])
results

Empty DataFrame
Columns: [accuracy, precison, recall, f1_score]
Index: []

In [210]:
results.loc['Logistic Regression'] = [accuracy,precision,recall , f1score]
results

accuracy  precison    recall  f1_score
Logistic Regression  0.768707  0.376471  0.680851  0.484848

In [211]:
# Implementation of Decision Tree Classifier

dt = DecisionTreeClassifier(criterion='gini',max_depth= 10,min_samples_split=10,min_samples_leaf=5,random_state = 10,max_leaf_nodes=10)
dt.fit(xtrain_res,ytrain_res)
print('Train accuracy = ',dt.score(xtrain_res,ytrain_res))
print('Test accuracy = ' ,dt.score(xtest,ytest))

Train accuracy =  0.8580121703853956
Test accuracy =  0.8163265306122449


In [212]:
# Checking out other metrics
ytest_pred = dt.predict(xtest)
accuracy = accuracy_score(y_true = ytest , y_pred = ytest_pred)
precision = precision_score(y_true = ytest , y_pred = ytest_pred)
recall = recall_score(y_true = ytest , y_pred = ytest_pred)
f1score = f1_score(y_true=ytest , y_pred = ytest_pred)
print('Accuracy Score  = ', accuracy)
print('Precision Score  = ', precision)
print('Recall Score = ' , recall)
print('F1 score = ' , f1score)

Accuracy Score  =  0.8163265306122449
Precision Score  =  0.4146341463414634
Recall Score =  0.3617021276595745
F1 score =  0.38636363636363635


In [213]:
# creating the confusion matrix
predicted = pd.Series(ytest_pred,name = 'predicted')
actual = pd.Series(ytest , name = 'actual')
pd.crosstab(index = predicted , columns = actual)


actual       0   1
predicted         
0          223  30
1           24  17

In [214]:
results.loc['Decision Tree(tuned)'] = [accuracy,precision,recall,f1score]
results

accuracy  precison    recall  f1_score
Logistic Regression   0.768707  0.376471  0.680851  0.484848
Decision Tree(tuned)  0.816327  0.414634  0.361702  0.386364

In [215]:
# Decision Tree without any hyperparameter tuning .

dt1 = DecisionTreeClassifier()
dt1.fit(xtrain_res,ytrain_res)
print('Train accuracy = ',dt1.score(xtrain_res,ytrain_res))
print('Test accuracy = ' ,dt1.score(xtest,ytest))
# Checking out other metrics
ytest_pred = dt1.predict(xtest)
accuracy = accuracy_score(y_true = ytest , y_pred = ytest_pred)
precision = precision_score(y_true = ytest , y_pred = ytest_pred)
recall = recall_score(y_true = ytest , y_pred = ytest_pred)
f1score = f1_score(y_true=ytest , y_pred = ytest_pred)
print('Accuracy Score  = ', accuracy)
print('Precision Score  = ', precision)
print('Recall Score = ' , recall)
print('F1 score = ' , f1score)

Train accuracy =  1.0
Test accuracy =  0.782312925170068
Accuracy Score  =  0.782312925170068
Precision Score  =  0.34545454545454546
Recall Score =  0.40425531914893614
F1 score =  0.37254901960784315


In [216]:
# creating the confusion matrix
predicted = pd.Series(ytest_pred,name = 'predicted')
actual = pd.Series(ytest , name = 'actual')
pd.crosstab(index = predicted , columns = actual)


actual       0   1
predicted         
0          211  28
1           36  19

In [217]:
results.loc['Decision Tree (no tuning)'] = [accuracy,precision,recall,f1score]
results

accuracy  precison    recall  f1_score
Logistic Regression        0.768707  0.376471  0.680851  0.484848
Decision Tree(tuned)       0.816327  0.414634  0.361702  0.386364
Decision Tree (no tuning)  0.782313  0.345455  0.404255  0.372549

In [218]:
# Random forest Implementation 

rf = RandomForestClassifier()
rf.fit(xtrain_res,ytrain_res)
print('Train accuracy = ',rf.score(xtrain_res,ytrain_res))
print('Test accuracy = ' ,rf.score(xtest,ytest))
# Checking out other metrics
ytest_pred = rf.predict(xtest)
accuracy = accuracy_score(y_true = ytest , y_pred = ytest_pred)
precision = precision_score(y_true = ytest , y_pred = ytest_pred)
recall = recall_score(y_true = ytest , y_pred = ytest_pred)
f1score = f1_score(y_true=ytest , y_pred = ytest_pred)
print('Accuracy Score  = ', accuracy)
print('Precision Score  = ', precision)
print('Recall Score = ' , recall)
print('F1 score = ' , f1score)

Train accuracy =  1.0
Test accuracy =  0.8639455782312925
Accuracy Score  =  0.8639455782312925
Precision Score  =  0.7333333333333333
Recall Score =  0.23404255319148937
F1 score =  0.3548387096774194


In [219]:
# creating the confusion matrix
predicted = pd.Series(ytest_pred,name = 'predicted')
actual = pd.Series(ytest , name = 'actual')
pd.crosstab(index = predicted , columns = actual)


actual       0   1
predicted         
0          243  36
1            4  11

In [220]:
results.loc['Random Forest (no tuning)'] = [accuracy,precision,recall,f1score]
results

accuracy  precison    recall  f1_score
Logistic Regression        0.768707  0.376471  0.680851  0.484848
Decision Tree(tuned)       0.816327  0.414634  0.361702  0.386364
Decision Tree (no tuning)  0.782313  0.345455  0.404255  0.372549
Random Forest (no tuning)  0.863946  0.733333  0.234043  0.354839

In [221]:
# Hyperparameter tuning of Random Forest classifier.
from sklearn.model_selection import GridSearchCV
params = {'n_estimators':[10,50,100],
          'max_depth':[10,20,30],
          'max_samples' : [500,1000,1500],
          'max_features' : ['sqrt'],
          'bootstrap' : [True],
          'criterion' : ['gini'],
          'min_samples_split':[10,20,30],
          'min_samples_leaf' : [10,20,30],
          'verbose' : [1],
          'n_jobs' : [-1]
          }
gcv = GridSearchCV(estimator=RandomForestClassifier(random_state=10),param_grid=params)
gcv.fit(xtrain_res,ytrain_res)

[Parallel(n_jobs=-1)]: Using backend ThreadingBackend with 16 concurrent workers.
[Parallel(n_jobs=-1)]: Done  10 out of  10 | elapsed:    0.0s finished
[Parallel(n_jobs=10)]: Using backend ThreadingBackend with 10 concurrent workers.
[Parallel(n_jobs=10)]: Done   2 out of  10 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=10)]: Done  10 out of  10 | elapsed:    0.0s finished
[Parallel(n_jobs=-1)]: Using backend ThreadingBackend with 16 concurrent workers.
[Parallel(n_jobs=-1)]: Done  10 out of  10 | elapsed:    0.0s finished
[Parallel(n_jobs=10)]: Using backend ThreadingBackend with 10 concurrent workers.
[Parallel(n_jobs=10)]: Done   2 out of  10 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=10)]: Done  10 out of  10 | elapsed:    0.0s finished
[Parallel(n_jobs=-1)]: Using backend ThreadingBackend with 16 concurrent workers.
[Parallel(n_jobs=-1)]: Done  10 out of  10 | elapsed:    0.0s finished
[Parallel(n_jobs=10)]: Using backend ThreadingBackend with 10 concurrent 

[Parallel(n_jobs=-1)]: Using backend ThreadingBackend with 16 concurrent workers.
[Parallel(n_jobs=-1)]: Done  18 tasks      | elapsed:    0.0s
[Parallel(n_jobs=-1)]: Done 100 out of 100 | elapsed:    0.0s finished
[Parallel(n_jobs=16)]: Using backend ThreadingBackend with 16 concurrent workers.
[Parallel(n_jobs=16)]: Done  18 tasks      | elapsed:    0.0s
[Parallel(n_jobs=16)]: Done 100 out of 100 | elapsed:    0.0s finished
[Parallel(n_jobs=-1)]: Using backend ThreadingBackend with 16 concurrent workers.
[Parallel(n_jobs=-1)]: Done  10 out of  10 | elapsed:    0.0s finished
[Parallel(n_jobs=10)]: Using backend ThreadingBackend with 10 concurrent workers.
[Parallel(n_jobs=10)]: Done   2 out of  10 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=10)]: Done  10 out of  10 | elapsed:    0.0s finished
[Parallel(n_jobs=-1)]: Using backend ThreadingBackend with 16 concurrent workers.
[Parallel(n_jobs=-1)]: Done  10 out of  10 | elapsed:    0.0s finished
[Parallel(n_jobs=10)]: Using b

GridSearchCV(estimator=RandomForestClassifier(random_state=10),
             param_grid={'bootstrap': [True], 'criterion': ['gini'],
                         'max_depth': [10, 20, 30], 'max_features': ['sqrt'],
                         'max_samples': [500, 1000, 1500],
                         'min_samples_leaf': [10, 20, 30],
                         'min_samples_split': [10, 20, 30],
                         'n_estimators': [10, 50, 100], 'n_jobs': [-1],
                         'verbose': [1]})

In [222]:
# Get the best parameters and the best score
best_params = gcv.best_params_
best_score = gcv.best_score_

print(f"Best parameters: {best_params}")
print(f"Best score: {best_score}")

Best parameters: {'bootstrap': True, 'criterion': 'gini', 'max_depth': 20, 'max_features': 'sqrt', 'max_samples': 1500, 'min_samples_leaf': 10, 'min_samples_split': 10, 'n_estimators': 100, 'n_jobs': -1, 'verbose': 1}
Best score: 0.9037165071001736


In [223]:
# Checking out other metrics
ytest_pred = gcv.predict(xtest)
accuracy = accuracy_score(y_true = ytest , y_pred = ytest_pred)
precision = precision_score(y_true = ytest , y_pred = ytest_pred)
recall = recall_score(y_true = ytest , y_pred = ytest_pred)
f1score = f1_score(y_true=ytest , y_pred = ytest_pred)
print('Accuracy Score  = ', accuracy)
print('Precision Score  = ', precision)
print('Recall Score = ' , recall)
print('F1 score = ' , f1score)

Accuracy Score  =  0.8469387755102041
Precision Score  =  0.5357142857142857
Recall Score =  0.3191489361702128
F1 score =  0.4


[Parallel(n_jobs=16)]: Using backend ThreadingBackend with 16 concurrent workers.
[Parallel(n_jobs=16)]: Done  18 tasks      | elapsed:    0.0s
[Parallel(n_jobs=16)]: Done 100 out of 100 | elapsed:    0.0s finished


In [224]:
# creating the confusion matrix
predicted = pd.Series(ytest_pred,name = 'predicted')
actual = pd.Series(ytest , name = 'actual')
pd.crosstab(index = predicted , columns = actual)


actual       0   1
predicted         
0          234  32
1           13  15

In [225]:
results.loc['Random Forest(tuned)'] = [accuracy,precision,recall,f1score]
results

accuracy  precison    recall  f1_score
Logistic Regression        0.768707  0.376471  0.680851  0.484848
Decision Tree(tuned)       0.816327  0.414634  0.361702  0.386364
Decision Tree (no tuning)  0.782313  0.345455  0.404255  0.372549
Random Forest (no tuning)  0.863946  0.733333  0.234043  0.354839
Random Forest(tuned)       0.846939  0.535714  0.319149  0.400000

In [226]:
# Logistic regression with grid search cv
lr = LogisticRegression(random_state = 10)
params = {'penalty':['None','l2','l1'] , 'tol':[0.001,0.0001,1],'C':[0.5,0.7,0.8],'max_iter':[100,150,200]}
gcv1 = GridSearchCV(estimator=lr,param_grid=params,n_jobs = -1)
gcv1.fit(xtrain_res,ytrain_res)

GridSearchCV(estimator=LogisticRegression(random_state=10), n_jobs=-1,
             param_grid={'C': [0.5, 0.7, 0.8], 'max_iter': [100, 150, 200],
                         'penalty': ['None', 'l2', 'l1'],
                         'tol': [0.001, 0.0001, 1]})

In [227]:
# Get the best parameters and the best score
best_params = gcv1.best_params_
best_score = gcv1.best_score_

print(f"Best parameters: {best_params}")
print(f"Best score: {best_score}")

Best parameters: {'C': 0.8, 'max_iter': 100, 'penalty': 'l2', 'tol': 0.001}
Best score: 0.8098348647433016


In [228]:
# Checking out other metrics
ytest_pred = gcv1.predict(xtest)
accuracy = accuracy_score(y_true = ytest , y_pred = ytest_pred)
precision = precision_score(y_true = ytest , y_pred = ytest_pred)
recall = recall_score(y_true = ytest , y_pred = ytest_pred)
f1score = f1_score(y_true=ytest , y_pred = ytest_pred)
print('Accuracy Score  = ', accuracy)
print('Precision Score  = ', precision)
print('Recall Score = ' , recall)
print('F1 score = ' , f1score)

Accuracy Score  =  0.7687074829931972
Precision Score  =  0.3793103448275862
Recall Score =  0.7021276595744681
F1 score =  0.4925373134328358


In [229]:
# creating the confusion matrix
predicted = pd.Series(ytest_pred,name = 'predicted')
actual = pd.Series(ytest , name = 'actual')
pd.crosstab(index = predicted , columns = actual)

actual       0   1
predicted         
0          193  14
1           54  33

In [230]:
results.loc['Logistic Regression(tuned)'] = [accuracy,precision,recall,f1score]


In [231]:
results

accuracy  precison    recall  f1_score
Logistic Regression         0.768707  0.376471  0.680851  0.484848
Decision Tree(tuned)        0.816327  0.414634  0.361702  0.386364
Decision Tree (no tuning)   0.782313  0.345455  0.404255  0.372549
Random Forest (no tuning)   0.863946  0.733333  0.234043  0.354839
Random Forest(tuned)        0.846939  0.535714  0.319149  0.400000
Logistic Regression(tuned)  0.768707  0.379310  0.702128  0.492537

In [232]:

results.sort_values(by = 'recall' , ascending=False)

accuracy  precison    recall  f1_score
Logistic Regression(tuned)  0.768707  0.379310  0.702128  0.492537
Logistic Regression         0.768707  0.376471  0.680851  0.484848
Decision Tree (no tuning)   0.782313  0.345455  0.404255  0.372549
Decision Tree(tuned)        0.816327  0.414634  0.361702  0.386364
Random Forest(tuned)        0.846939  0.535714  0.319149  0.400000
Random Forest (no tuning)   0.863946  0.733333  0.234043  0.354839

In [233]:
# Precision :  “Of all the instances classified as positive, how many are actually positive?” A high precision indicates a low rate of false positives.


# Recall : “Of all the actual positives, how many were identified correctly?” A high recall indicates a low rate of false negatives.

## Conclusion


### In This Employee Attrition Likelihood Prediction Model , it is most important for our model to correctly predict the True Positives i.e the employees who are likely to leave the company . In other words , the model with the highest recall score is the most appropriate model . We need less False negatives in the model . 

### Hence looking at the Recall scores of the above models , We see that the tuned logistic Regression ( Tuned) is performing the best out of all with the Recall score of 70.21%  .

### Logistic regression (Tuned) model is giving a higher False positive instances compared to the DT model and Random Forest models.The best precision or the lowest false positive is given by the Random Forest models , however there recall scores are lowest hence they cannot be chosen as the model for this particular project.

### As per the Confusion Matrix of Logistic regression (tuned) we can see that the False Negative instances = 14 and False Positive instances = 54.The True Positive instances = 33 (the most significant value) .